### PR TITLE
Prevent modification of value in `INTENT(IN)`

### LINKED ISSUE #1425: `INTENT(IN)` should not allow modification of values
I was looking into data types (immutable vs mutable) in Fortran and came across this issue.

```
PROGRAM immutable_example
  INTEGER :: x
  x = 42
  CALL try_to_change(x)
  WRITE(*,*) "The value of x is", x
END PROGRAM

SUBROUTINE try_to_change(y)
  INTEGER, INTENT(IN) :: y
  y = 99  
END SUBROUTINE

(lf) faze-geek@DESKTOP-497JDCU:~/lfortran/lfortran$ gfortran try.f90  && ./a.out
try.f90:10:2:

   10 |   y = 99
      |  1
Error: Dummy argument ‘y’ with INTENT(IN) in variable definition context (assignment) at (1)

(lf) faze-geek@DESKTOP-497JDCU:~/lfortran/lfortran$ ./src/bin/lfortran try.f90
The value of x is 99
```
Modification the value of an argument that is declared with` INTENT(IN)` is not allowed.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1425.

### CODE REVIEW

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 1]
```diff
@@ -1789,6 +1789,13 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                 std::string var_name = std::string(v->m_name);
                 throw SemanticError("Assignment to loop variable `" + std::string(to_lower(var_name)) +"` is not allowed", target->base.loc);
             }
+            if (sym->type == ASR::symbolType::Variable) {
+                ASR::Variable_t *v = ASR::down_cast<ASR::Variable_t>(sym);
+                ASR::intentType intent = v->m_intent;
+                if (intent == ASR::intentType::In) {
+                    throw SemanticError("Argument `" + std::string(v->m_name) + "` with INTENT(IN) in variable definition context (assignment)", target->base.loc);
+                }
+            }
```
Comment (czgdp1807) [inline src/lfortran/semantics/ast_body_visitor.cpp:1798]: For long term: I feel these kinds of checks are comprised of two parts, "When?" and "Why?".

When? - This part detects when the assignment to a certain variable should be prohibited.
Why? - This part is the error message explaining why assignment is not allowed.

I feel this can be abstracted into a function which will do both When and Why for us so that we have to extend only that function.

P.S. Approving as the above is just an idea for long term.

**Review by czgdp1807 — APPROVED** [review]

File: src/lfortran/semantics/ast_body_visitor.cpp (hunk 1 above)
Comment (Pranavchiku) [reply src/lfortran/semantics/ast_body_visitor.cpp:1798]: yeah, along with that we stop execution of code as we receive an error, sometimes code has multiple errors and all are shown by GFortran with time we need to update that as well.

Comment (czgdp1807) [reply src/lfortran/semantics/ast_body_visitor.cpp:1798]: It depends if we can show multiple errors all at once. If a certain error doesn’t allow further AST to ASR conversion of code then we will have to stop compilation. Otherwise keep adding errors and show them together.

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_body_visitor.cpp:1798]: Agreed, in cases like this PR we can go ahead and show multiple errors at once. This might not be the case everytime though.

File: tests/reference/asr-intent1-4eee5ff.stderr [hunk 2]
```diff
@@ -0,0 +1,5 @@
+semantic error: Argument `y` with INTENT(IN) in variable definition context (assignment)
```
Comment (certik) [inline tests/reference/asr-intent1-4eee5ff.stderr:1]: Maybe rephrasing it to `Cannot assign to an intent(in) variable 'y'`?

Comment (certik) [reply tests/reference/asr-intent1-4eee5ff.stderr:1]: I feel using simple English and everyday terms helps with the clarity. We should still be precise, but I think the above is still as precise, but easier to understand.

Comment (Pranavchiku) [reply tests/reference/asr-intent1-4eee5ff.stderr:1]: Actually I copied over the error message from GFortran.

Comment (certik) [reply tests/reference/asr-intent1-4eee5ff.stderr:1]: I know you did. I don't think this message is very clear and user friendly in GFortran.

Comment (Pranavchiku) [reply tests/reference/asr-intent1-4eee5ff.stderr:1]: Okay, I will rephrase it to `Cannot assign to an intent(in) variable 'y'`
